### ISSUE TITLE
Consumer not consuming multiple partitions at the same speed

### ISSUE THREAD

**Issue by champtar** [edenhill/librdkafka#1395]
Description
===========
3 brokers / 6 partitions / 3 consumers using librdkafka 0.11 (2 partitions per consumer)
no news logs, only the consumers consuming
My application was reading at around ~5000msg/s, lag of 0 / 32817413 on one partition and 15248533 /  32824216 on the other (for one of the consumer)
```
TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID                                       HOST                           CLIENT-ID
vme1                           0          32817413        32817413        0          analytics1-29e59b52-0612-4ade-b36e-5a458d88f14c   /192.168.1.151                 analytics1
vme1                           1          17575683        32824216        15248533   analytics1-29e59b52-0612-4ade-b36e-5a458d88f14c   /192.168.1.151                 analytics1
...
```

I expect the partition to be read at about the same speed, not one after the other

Message are produced using kafkacat with librdkafka 0.11 (msgv2 / lz4)
```kafkacat -P -E -X compression.codec=lz4 -X queue.buffering.max.ms=100 -X security.protocol=sasl_plaintext -X sasl.mechanisms=PLAIN -X sasl.username=kafkacat -X sasl.password='...' -b ... -t vme1```

The setup is 3 VMs with broker + consumer on 1 host (so no network issues I think)

How to reproduce
================
No idea, what logs should I enable when I try to reproduce it ? (this is a lab setup so I can test everything you ask)

Checklist
=========
Please provide the following information:

 - [X] librdkafka version (release number or git tag): a212ea2 (was 0.11)
 - [X] Apache Kafka version: 0.11.0.1 (scala 2.11) (was 0.11.0.0)
 - [X] librdkafka client configuration: (consumer)
security.protocol = sasl_plaintext
sasl.mechanisms = PLAIN
api.version.request = true
api.version.fallback.ms = 0

 - [X] Operating system: Centos 7.4 (was 7.3)
 - [ ] Using the legacy Consumer
 - [X] Using the high-level KafkaConsumer
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
What's the typical message size?

**Comment by edenhill** [comment]
librdkafka will request all assigned partitions in a round-robin fashion, and the broker will serve those partitions in order, sending as many messages as possible per partition before moving on the next one, while at the same time being limited by `fetch.message.max.bytes`.

**Comment by champtar** [comment]
Hi @edenhill 

Just reproduced it with debug=all, here the log of consumer 1
https://drive.google.com/file/d/0BxXIIjPi2sGbMDhqMGZzdjcxZ2c/view?usp=sharing

```
[root@analytics3 ~]# /opt/kafka/bin/kafka-consumer-groups.sh --bootstrap-server 127.0.0.1:9092 --command-config="/etc/kafka/client.properties" --describe --group hindsight_group
Note: This will only show information about consumers that use the Java consumer API (non-ZooKeeper-based consumers).

TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID                                       HOST                           CLIENT-ID
vme1                           0          22099564        32828438        10728874   analytics1-3023a957-d5a1-4e84-90de-dee50f1c4270   /192.168.1.151                 analytics1
vme1                           1          32824568        32824568        0          analytics1-3023a957-d5a1-4e84-90de-dee50f1c4270   /192.168.1.151                 analytics1
vme1                           2          24852482        32818809        7966327    analytics2-62ce67d6-4167-430f-a4f6-7bed79070acf   /192.168.1.152                 analytics2
vme1                           3          26480141        32822935        6342794    analytics2-62ce67d6-4167-430f-a4f6-7bed79070acf   /192.168.1.152                 analytics2
vme1                           4          31779003        32823048        1044045    analytics3-5e5a3463-7d6c-44c6-937f-6d8937ba83bd   /192.168.1.153                 analytics3
vme1                           5          19831316        32819681        12988365   analytics3-5e5a3463-7d6c-44c6-937f-6d8937ba83bd   /192.168.1.153                 analytics3
[root@analytics3 ~]# date
mer. août 30 14:59:50 PDT 2017
```
Regarding the typical message size I'm sending as fast as possible log lines via tcp to socat which "pipe" it to kafkacat (ie kafkacat is my producer)
average log line size is 265.543 bytes, but maybe that's not what you are looking for?

```
[thrd:sasl_plaintext://192.168.1.151:9092/bootstrap]: sasl_plaintext://192.168.1.151:9092/1: Topic vme1 [0] MessageSet size 1048576, error "Success", MaxOffset 53096, Ver 6/6
[thrd:sasl_plaintext://192.168.1.151:9092/bootstrap]: sasl_plaintext://192.168.1.151:9092/1: Enqueue 3616 message(s) on vme1 [0] fetch queue (qlen 3616, v6, last_offset 31578)
```

**Comment by edenhill** [comment]
Ah, I think you are hitting this issue that was identified today: #1384 
I'll push a fix for it later today and I'd like you to try it out then.

**Comment by champtar** [comment]
Thanks, testing right now!

**Comment by champtar** [comment]
Hi @edenhill 

I add a new occurrence today, so not fixed :(
```
TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID                                       HOST                           CLIENT-ID
vme1                           4          29427752        32825985        3398233    analytics3-in-vme1-c8551217-a11f-4637-99d0-193a1d97ca7d/192.168.1.153                 analytics3-in-vme1
vme1                           5          23098314        32826668        9728354    analytics3-in-vme1-c8551217-a11f-4637-99d0-193a1d97ca7d/192.168.1.153                 analytics3-in-vme1
vme1                           0          32816650        32816650        0          analytics1-in-vme1-2e6cb3de-318c-4818-82b1-d5a1336462b2/192.168.1.151                 analytics1-in-vme1
vme1                           1          19988890        32821306        12832416   analytics1-in-vme1-2e6cb3de-318c-4818-82b1-d5a1336462b2/192.168.1.151                 analytics1-in-vme1
vme1                           2          19988647        32812880        12824233   analytics2-in-vme1-8319325e-fa29-4790-b8d3-25eecba9c4db/192.168.1.152                 analytics2-in-vme1
vme1                           3          32817071        32817071        0          analytics2-in-vme1-8319325e-fa29-4790-b8d3-25eecba9c4db/192.168.1.152                 analytics2-in-vme1
```

I'm now using kafka 0.11.0.1 / librdkafka 0.11.0.git (a212ea2) / CentOS 7.4

Tell me what logs you need
Thanks

**Comment by edenhill** [comment]
Looking at the original log in https://github.com/edenhill/librdkafka/issues/1395#issuecomment-326138696 which is for the first consumer consuming partitions 0 and 1, with a supposed lag of partition 0.

The kafka-consumer-groups tools bases the lag calculation on committed offsets, and looking at the logs of the time it was run (14:59) we see that the consumer is still trying to catch up to partition 0:
```
août 30 14:59:00 analytics1 hindsight[26380]: 1504130340137783033 [debug] analytics1#consumer-1 OFFSET        [thrd:main]: Topic vme1 [0]: stored offset 22099217, committed offset 22099167
août 30 14:59:00 analytics1 hindsight[26380]: 1504130340137815436 [debug] analytics1#consumer-1 OFFSET        [thrd:main]: Topic vme1 [0]: setting stored offset 22099217 for commit
août 30 14:59:00 analytics1 hindsight[26380]: 1504130340137819448 [debug] analytics1#consumer-1 OFFSET        [thrd:main]: Topic vme1 [1]: stored offset 32824568, committed offset 32824568
août 30 14:59:00 analytics1 hindsight[26380]: 1504130340137823271 [debug] analytics1#consumer-1 OFFSET        [thrd:main]: Topic vme1 [1]: setting offset INVALID for commit
````

26 minutes later it finally catches up:

```
août 30 15:25:55 analytics1 hindsight[26380]: 1504131955262400928 [debug] analytics1#consumer-1 OFFSET        [thrd:main]: Topic vme1 [0]: stored offset 32828438, committed offset 32725769
août 30 15:25:55 analytics1 hindsight[26380]: 1504131955262438168 [debug] analytics1#consumer-1 OFFSET        [thrd:main]: Topic vme1 [0]: setting stored offset 32828438 for commit
août 30 15:25:55 analytics1 hindsight[26380]: 1504131955262447148 [debug] analytics1#consumer-1 OFFSET        [thrd:main]: Topic vme1 [1]: stored offset 32824568, committed offset 32824568
août 30 15:25:55 analytics1 hindsight[26380]: 1504131955262455655 [debug] analytics1#consumer-1 OFFSET        [thrd:main]: Topic vme1 [1]: setting offset INVALID for commit
août 30 15:25:55 analytics1 hindsight[26380]: 1504131955262475372 [debug] analytics1#consumer-1 OFFSET        [thrd:main]: sasl_plaintext://192.168.1.152:9092/2: Enqueue OffsetCommitRequest(v1, 1/2 partition(s))): cgrp auto commit timer
août 30 15:25:55 analytics1 hindsight[26380]: 1504131955262655025 [debug] analytics1#consumer-1 SEND        [thrd:sasl_plaintext://192.168.1.152:9092/bootstrap]: sasl_plaintext://192.168.1.152:9092/2: Sent OffsetCommitRequest (v1, 130 bytes @ 0, CorrId 32633)
...
août 30 15:25:55 analytics1 hindsight[26380]: 1504131955319028673 [debug] analytics1#consumer-1 RECV        [thrd:sasl_plaintext://192.168.1.152:9092/bootstrap]: sasl_plaintext://192.168.1.152:9092/2: Received OffsetCommitResponse (v1, 20 bytes, CorrId 32633, rtt 56.35ms)
août 30 15:25:55 analytics1 hindsight[26380]: 1504131955319377980 [debug] analytics1#consumer-1 COMMIT        [thrd:main]: sasl_plaintext://192.168.1.152:9092/2: OffsetCommit for 2 partition(s): cgrp auto commit timer: returned: Success
```

The pro-longed catchup could be caused by the previously mentioned librdkafka bug (fixed on master), slow network, or slow processing.

With the new occurrence of this issue, does it eventually catch up if you wait?

**Comment by champtar** [comment]
My test process is:
clean everything (stop kafka / rm kafka data dir / start kafka)
produce a big batch of logs with kafkacat (197 millions log lines) at ~150/200k lines per sec
consume the logs from 3 clients at ~15/20k per sec (total)

The partition are not consumed at the same pace but yes in the end everything is consumed,
what I'm worried about is if I continue to feed logs will the consumers continue to consume first on 1 partition (i will try to test that but my test environment is pretty small sadly)

**Comment by champtar** [comment]
Just tried to go through the logs, found some PROTOERR, but they seem unrelated
I made some rough graphs of the ```last_offset``` of the ```CONSUME``` logs, and I just see that one of the partition is consumed ~2 time faster than the other, but i don't see why
New logs (the 3 consumers): https://drive.google.com/drive/folders/0BxXIIjPi2sGbUWdVMkRKajVzalU?usp=sharing

Quick question to help me understand, there is 1 message queue per consumer, per topic or per partition? (queued.min.messages/queued.max.messages.kbytes)
I see a lot of ```FETCHADD``` after ```FETCH ... is not fetchable: queued.min.messages exceeded``` (looking at what can introduce "unfairness", without any luck)

**Comment by champtar** [comment]
(can you reopen the issue?)

**Comment by champtar** [comment]
ping @edenhill, can you reopen this issue so it's not forgotten?

**Comment by edenhill** [comment]
By default, and legacy, there is one fetchq per partition, but when using the rd_kafka_poll_set_consumer() API the partition queues are forwarded to one single consumer queue, and this shared queue is subject to queued.min.messages and queued.max.message.kbytes.

**Comment by edenhill** [comment]
Can you tell me what to look for in the new logs: what topic+partition is lagging and which is not lagging?

**Comment by champtar** [comment]
Hi @edenhill

There is only 1 topic (vme), and for example partition 3 is consumed ~ 2 times faster than partition 2
(i'm trying to show that with the previous graphs but my LibreOffice skills are failing me ;) )
relevant file is analytics2.logs.xz

This is a lab and I reproduce it almost every time, so I can run any modified librdkafka if needed
The client code is https://github.com/mozilla-services/lua_sandbox_extensions/blob/master/kafka/kafka.c

Thanks

**Comment by edenhill** [comment]
Sorry for dropping the ball on this one.
Is this still an issue?

**Comment by champtar** [comment]
Hi @edenhill 
Last time I checked it was still happening, but I haven't tried the very last version of librdkafka yet, will try to redo tests tomorrow

**Comment by champtar** [comment]
Hi @edenhill 

Issue is still happening with latest librdkafka (cb6f08c):
```
TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID                                       HOST                           CLIENT-ID
vme1                           4          8193058         32823521        24630463   analytics3-in-vme1-cf5d8421-fc42-4398-9b17-4a942787496f/192.168.1.153                 analytics3-in-vme1
vme1                           5          6796458         32821245        26024787   analytics3-in-vme1-cf5d8421-fc42-4398-9b17-4a942787496f/192.168.1.153                 analytics3-in-vme1
vme1                           2          4807886         32825163        28017277   analytics2-in-vme1-d7416283-def5-44e9-ad88-ea13f44ef062/192.168.1.152                 analytics2-in-vme1
vme1                           3          10539411        32820536        22281125   analytics2-in-vme1-d7416283-def5-44e9-ad88-ea13f44ef062/192.168.1.152                 analytics2-in-vme1
vme1                           0          9976704         32807449        22830745   analytics1-in-vme1-6c564b11-13df-451b-918b-8be3a23cc27a/192.168.1.151                 analytics1-in-vme1
vme1                           1          4368358         32822646        28454288   analytics1-in-vme1-6c564b11-13df-451b-918b-8be3a23cc27a/192.168.1.151                 analytics1-in-vme1
```

**Comment by champtar** [comment]
Some time later
```
TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID                                       HOST                           CLIENT-ID
vme1                           4          26574360        32823521        6249161    analytics3-in-vme1-cf5d8421-fc42-4398-9b17-4a942787496f/192.168.1.153                 analytics3-in-vme1
vme1                           5          21974211        32821245        10847034   analytics3-in-vme1-cf5d8421-fc42-4398-9b17-4a942787496f/192.168.1.153                 analytics3-in-vme1
vme1                           2          14967417        32825163        17857746   analytics2-in-vme1-d7416283-def5-44e9-ad88-ea13f44ef062/192.168.1.152                 analytics2-in-vme1
vme1                           3          32820536        32820536        0          analytics2-in-vme1-d7416283-def5-44e9-ad88-ea13f44ef062/192.168.1.152                 analytics2-in-vme1
vme1                           0          32807449        32807449        0          analytics1-in-vme1-6c564b11-13df-451b-918b-8be3a23cc27a/192.168.1.151                 analytics1-in-vme1
vme1                           1          14980434        32822646        17842212   analytics1-in-vme1-6c564b11-13df-451b-918b-8be3a23cc27a/192.168.1.151                 analytics1-in-vme1
```

**Comment by edenhill** [comment]
Looking through the logs the only difference I can see between partition 3 and 2 is that they are on different brokers:
 * partition 2 is on broker 1
 * partition 3 is on broker 2

It turns out that broker 1 is also the group coordinator, which means that offset commits for both partitions will be sent to this broker, and since a broker only processes on request (be it produce, fetch or commit) at the time this also means that every now and then - every commit interval - the Fetch request behind the Commit request will be stalled until the commit succeeds.

My guess is that if you looked at the per-broker rtt stats (https://github.com/edenhill/librdkafka/wiki/Statistics#brokers) you would see that broker 1 would have a higher avg/max RTT and a slightly lower tx/rx rate.

This is an area where the Java and librdkafka consumers differ:
the Java consumer uses a dedicated connection for offset commits while librdkafka uses one single connection per broker and all requests are using this same connection, thus ending up with this head of line blocking.

Workarounds for the time being is to increase the auto commit interval.

**Comment by edenhill** [comment]
This issue is now fixed on master

**Comment by champtar** [comment]
Hey @edenhill 
Sadly it doesn't seems fixed for me (test using ce67923)
```
/opt/kafka/bin/kafka-consumer-groups.sh --bootstrap-server 127.0.0.1:9092 --command-config="/etc/kafka/client.properties" --describe --group hindsight_group_vme1

TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID                                       HOST                           CLIENT-ID
vme1                           0          24980666        32820269        7839603    analytics1-in-vme1-cfd428a2-3bac-4417-8367-ae68c9cdd76c/192.168.1.151                 analytics1-in-vme1
vme1                           1          15809489        32817525        17008036   analytics1-in-vme1-cfd428a2-3bac-4417-8367-ae68c9cdd76c/192.168.1.151                 analytics1-in-vme1
vme1                           4          27108818        32821511        5712693    analytics3-in-vme1-16c661af-9dc7-4591-956b-3fb0fcdcfa00/192.168.1.153                 analytics3-in-vme1
vme1                           5          18924089        32819954        13895865   analytics3-in-vme1-16c661af-9dc7-4591-956b-3fb0fcdcfa00/192.168.1.153                 analytics3-in-vme1
vme1                           2          13858328        32823749        18965421   analytics2-in-vme1-b9dae79e-e30b-4e18-9e4a-e08ae54fb0da/192.168.1.152                 analytics2-in-vme1
vme1                           3          27474530        32817552        5343022    analytics2-in-vme1-b9dae79e-e30b-4e18-9e4a-e08ae54fb0da/192.168.1.152                 analytics2-in-vme1
```
```
[root@analytics3 ~]# /opt/kafka/bin/kafka-topics.sh --zookeeper 127.0.0.1:2181 --describe --topic vme1
Topic:vme1	PartitionCount:6	ReplicationFactor:2	Configs:
	Topic: vme1	Partition: 0	Leader: 2	Replicas: 2,3	Isr: 3,2
	Topic: vme1	Partition: 1	Leader: 3	Replicas: 3,1	Isr: 3,1
	Topic: vme1	Partition: 2	Leader: 1	Replicas: 1,2	Isr: 1,2
	Topic: vme1	Partition: 3	Leader: 2	Replicas: 2,1	Isr: 1,2
	Topic: vme1	Partition: 4	Leader: 3	Replicas: 3,2	Isr: 3,2
	Topic: vme1	Partition: 5	Leader: 1	Replicas: 1,3	Isr: 3,1
```

i'm expecting fair round robin between partitions (I know kafka doesn't guarantee that)

**Comment by edenhill** [comment]
@champtar ack. Can you reproduce with debug=msg and provide a log excerpt covering both fast and slow partitions?

**Comment by champtar** [comment]
@edenhill I will (but not today)

**Comment by champtar** [comment]
Still happening with librdkafka 0.11.4 & kafka 1.1.0
I'll reproduce with debug then we can reopen
```
[root@analytics3 ~]# /opt/kafka/bin/kafka-consumer-groups.sh --bootstrap-server 127.0.0.1:9092 --command-config="/etc/kafka/client.properties" --describe --group hindsight_group_vme1
Note: This will not show information about old Zookeeper-based consumers.

TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID                                             HOST            CLIENT-ID
vme1            0          32828999        32828999        0               analytics1-in-vme1-595403f0-edec-4311-b308-4f3e2033f313 /192.168.1.151  analytics1-in-vme1
vme1            1          21561694        32822280        11260586        analytics1-in-vme1-595403f0-edec-4311-b308-4f3e2033f313 /192.168.1.151  analytics1-in-vme1
vme1            2          32656299        32809506        153207          analytics2-in-vme1-701f5c83-f678-48d0-bd1a-d8fa898ba36c /192.168.1.152  analytics2-in-vme1
vme1            3          32825060        32825060        0               analytics2-in-vme1-701f5c83-f678-48d0-bd1a-d8fa898ba36c /192.168.1.152  analytics2-in-vme1
vme1            4          31771140        32812964        1041824         analytics3-in-vme1-71b80d00-db53-4b64-81ca-5e43aa360e51 /192.168.1.153  analytics3-in-vme1
vme1            5          22538990        32821751        10282761        analytics3-in-vme1-71b80d00-db53-4b64-81ca-5e43aa360e51 /192.168.1.153  analytics3-in-vme1
```

**Comment by champtar** [comment]
Here the logs of hindsight (librdkafka consumer) with debug=msg
https://drive.google.com/drive/folders/0BxXIIjPi2sGbUWdVMkRKajVzalU?usp=sharing
(20180410 folder)

```
[root@analytics1 ~]# /opt/kafka/bin/kafka-consumer-groups.sh --bootstrap-server 127.0.0.1:9092 --command-config="/etc/kafka/client.properties" --describe --group hindsight_group_vme1
Note: This will not show information about old Zookeeper-based consumers.
TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID                                             HOST            CLIENT-ID
vme1            2          14701777        32821308        18119531        analytics2-in-vme1-1a6f604b-d491-4cde-9cf3-e07b237a7087 /192.168.1.152  analytics2-in-vme1
vme1            3          28021468        32822299        4800831         analytics2-in-vme1-1a6f604b-d491-4cde-9cf3-e07b237a7087 /192.168.1.152  analytics2-in-vme1
vme1            0          23752004        32829209        9077205         analytics1-in-vme1-06226d7e-c4ae-49bc-bb42-be0c1eb7b9f8 /192.168.1.151  analytics1-in-vme1
vme1            1          14111821        32815536        18703715        analytics1-in-vme1-06226d7e-c4ae-49bc-bb42-be0c1eb7b9f8 /192.168.1.151  analytics1-in-vme1
vme1            4          23134824        32816834        9682010         analytics3-in-vme1-7bbb26d8-e1d4-4057-af91-bc277948476c /192.168.1.153  analytics3-in-vme1
vme1            5          15460724        32815374        17354650        analytics3-in-vme1-7bbb26d8-e1d4-4057-af91-bc277948476c /192.168.1.153  analytics3-in-vme1
```

**Comment by edenhill** [comment]
Looking at analytis3 it looks like it is fetching its 2 partitions at 100ms intervals, it would be interesting to know if that pacing is in librdkafka or the broker. To find out you would need to run with `debug=msg,protocol` to see per-request rtt latencies.

There's also a large number of large intervals between fetches that are suspiciously close to a multiple of 1000ms.
Such nice numbers are always suspicious..

**Comment by champtar** [comment]
Folder 20180411 contains logs with `debug=msg,protocol`
https://drive.google.com/drive/folders/0BxXIIjPi2sGbUWdVMkRKajVzalU?usp=sharing

I see a lot of PROTOUFLOW logs :(

Something I forgot to mention is that I have the following sysctls tunnings:
```
net.core.rmem_max=10000000
net.core.wmem_max=10000000
net.ipv4.tcp_rmem=4096 87380 16777216
net.ipv4.tcp_wmem=4096 16384 16777216
```

**Comment by edenhill** [comment]
PROTOUFLOW that only show up in debugging and says "(truncated response from broker (ok))" are okay and expected, the broker may truncate its response (due to sendfile())

**Comment by edenhill** [comment]
Look at the time between FetchResponse and the next FetchRequest it shows exactly 1s most of the time, so delay delay is in librdkafka.

I would now like you to reproduce with `debug=msg,protocol,fetch` (last one, I promise, maybe)

**Comment by champtar** [comment]
folder 20180412 ;)
thanks for all your work!

**Comment by edenhill** [comment]
The partition is decided "non fetchable" when the shared fetchq limit is reached, at which time the partition is not fetched again until `socket.blocking.max.ms` has passed (and sometimes more).
I think you should try two things:
 * first, increase queued.min.messages to allow a larger local fetch queue. I think a reasonable recommendation would be to have this value to allow for 3-5s of consumption (i.e., your max message processing rate * 5).
 * if that doesn't fully help, decrease socket.blocking.max.ms to some value between 1-100ms. The lower value the higher CPU usage.

**Comment by champtar** [comment]
Hi @edenhill 
Just to be sure to understand the behaviour, librdkafka try at the same time to fetch from 2 partitions that are on 2 differents brokers, one of the broker is faster to respond to the consumer, so we fetch from it, and we hit `queued.min.messages` so we don't fetch from the second one, and librdkafka even mark the partition as non fetchable which increase the unfainess ?

I'll try to play with `queued.min.messages`.
Thanks !!

**Comment by accelerated** [comment]
Hi @edenhill, I am seeing the same issue with the latest library version. I have a very simple setup with 1 broker, 1 topic and 4 partitions and one single consumer. I have a producer send off say 1000 messages (less than 50 bytes each) and then stop. The 4 partitions each have 250 messages queued up. When I run the consumer, I can see it reads all partitions in order but it reads them to completion before moving on to the next, no round-robin is happening at all. Is this behavior as-designed? I have not changed any `queued.min.messages` settings, everything is left at default values. If the producer keeps producing, I believe the consumer keeps reading from the same partition and never even reads the others.

**Comment by edenhill** [comment]
> I can see it reads all partitions in order but it reads them to completion before moving on to the next, no round-robin is happening at all. Is this behavior as-designed?

You mean the order which you receive the messages in your application?
Since messages are fetched in batches this makes sense, the FetchResponse from the broker will first contain messages for partition 0, then for partition 1, and so on, and they are enqueued to the application in that same order.

**Comment by accelerated** [comment]
@edenhill well my current setting for `batch.num.messages  = 1` so theoretically I'm just reading one message at a time. It's not the order in which i receive messages (yes they all come from the same partition) but I also see when I dump the consumer group partitions using the kafka shell script `kafka-consumer-groups.sh`, I can see on the broker that only one partition is being emptied out and it's emptied one message at a time which corresponds to the setting above. So I definitely think it's an issue inside the library.

**Comment by accelerated** [comment]
Ok I finally got it to working slightly better by reducing the fetch message size, in which case the broker thread will batch less messages from each partition, therefore allowing a better distribution. I was my misunderstanding that the global forward queue was _optionally_ enabled. Is there a way to completely turn off the global queue forwarding and read from individual partition queues while maintaining the auto-rebalance feature?

**Comment by edenhill** [comment]
The partition fetch queues are forwarded to a single consumer queue by default (to allow polling just a single queue for messages from all applications).
You can remove this forwarding by getting the partition queue with `rd_kafka_queue_get_partition()` and then calling `rd_kafka_queue_get_partition(the_part_queue, NULL);` to disable forwarding.
The partition queue may then be polled separately with `rd_kafka_consume_queue(the_part_queue, ..)`.
Do note that you still need to poll the consumer queue for rebalance events, etc. (it too can be redirected if needed).

**Comment by briot** [comment]
> The partition fetch queues are forwarded to a single consumer queue by default (to allow polling just a single queue for messages from all applications).
> You can remove this forwarding by getting the partition queue with rd_kafka_queue_get_partition() and then calling rd_kafka_queue_get_partition(the_part_queue, NULL); to disable forwarding.
> The partition queue may then be polled separately with rd_kafka_consume_queue(the_part_queue, ..).
> Do note that you still need to poll the consumer queue for rebalance events, etc. (it too can be redirected if needed).


Magnus,

This is very interesting documentation.
Would you accept pull requests for adding this kind of gems, taken from your own messages to github, to rdkafka.h
directly ?

Emmanuel

**Comment by edenhill** [comment]
@briot Absolutely!

**Comment by accelerated** [comment]
Hi and thanks for the quick response. I think it would be awesome if @edenhill could provide an extra section in the documentation regarding this very important and often missunderstod topic (i.e. all the polling scenarios and also which ones are deprecated). I nearly spent a whole day yesterday debugging how the partition forwarding works because I didn't know what happened under the hood and I was seeing partitions being consumed at different rates. And the solution to lower the batch max bytes size is not ideal either. I will test the above APIs today and hopefully this will allow full round-robin consumption.

PS: I assume that if we disable forwarding as indicated above, calling `rd_kafka_poll_set_consumer()` will be a no-op and we still have to explicitly call `rd_kafka_poll()` instead?

**Comment by rolandyoung** [comment]
My +1 for the documentation request. The Confluent documentation is also silent on the polling strategies available and there seem to be a very large number of rd_kafka_consume* functions, not always annotated to explain which API they belong to.

**Comment by accelerated** [comment]
Ok so I have implemented the above code changes to allow polling from multiple partitions, however when calling `rd_kafka_poll()`  I am not getting any events at all and since i'm relying on the partition assignment to get the other queues, nothing works. Is there something else I'm missing? (Note that I am _not_ calling `rd_kafka_poll_set_consumer()` at all to forward the main event queue so theoretically `rd_kafka_poll` should work. 
I also tried setting `rd_kafka_queue_forward(rd_kafka_queue_get_main(ptr), NULL)` and still nothing happens.
